{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "acde4e9f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/jonas/Documents/NTNU/Bachelor/code-model-embeddings/venv/lib/python3.14/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "sys.path.append(os.path.abspath(\"..\"))\n",
    "\n",
    "from data.dataset_factory import get_dataset_generator\n",
    "from data.data_generators.bigclonebench_dataset_gen import selected_generator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1d5b8466",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "import java.io.BufferedWriter;\n",
      "import java.io.FileWriter;\n",
      "import java.io.IOException;\n",
      "import java.io.Writer;\n",
      "import java.util.ArrayList;\n",
      "import cern.jet.random.Gamma;\n",
      "\n",
      "public class NaiveCalling {\n",
      "\n",
      "    static ArrayList<Exon> Exons = new ArrayList<Exon>();\n",
      "\n",
      "    static ArrayList<Exon> ExpectedValues = new ArrayList<Exon>();\n",
      "\n",
      "    static ArrayList<Exon> StdDeviations = new ArrayList<Exon>();\n",
      "\n",
      "    static String exonFileName = \"\";\n",
      "\n",
      "    static String outputFileName = \"\";\n",
      "\n",
      "    static ArrayList<String> baselineExonFileNames = new ArrayList<String>();\n",
      "\n",
      "    static ArrayList<Integer> chromosomes = new ArrayList<Integer>();\n",
      "\n",
      "    static String parameterFileName = \"\";\n",
      "\n",
      "    static boolean printTimingMetrics = false;\n",
      "\n",
      "    static boolean usingPileup = false;\n",
      "\n",
      "    static double threshold = 0;\n",
      "\n",
      "    HiddenMarkovModel hmm;\n",
      "\n",
      "    public static void main(String args[]) {\n",
      "        parseArguments(args);\n",
      "        try {\n",
      "            HiddenMarkovModel.initialize(parameterFileName);\n",
      "            int nSamples = baselineExonFileNames.size();\n",
      "            System.out.println(\"Number of baseline: \" + nSamples);\n",
      "            if (nSamples < 10) for (String s : baselineExonFileNames) System.out.println(s);\n",
      "            for (int chromosome : chromosomes) {\n",
      "                System.out.println(\"Calculating expected values from given files....\");\n",
      "                ExpectedValues = Exon.calculateExpectedValues(baselineExonFileNames, chromosome);\n",
      "                StdDeviations = Exon.calculateStdDevValues(baselineExonFileNames, chromosome, ExpectedValues);\n",
      "                Exon.sortExons(ExpectedValues);\n",
      "                Exon.sortExons(StdDeviations);\n",
      "                System.out.println(\"Reading exon with measurements file....\");\n",
      "                Exons = Exon.readAndStoreExonFile(exonFileName, chromosome);\n",
      "                Exon.sortExons(Exons);\n",
      "                double genomeWideAve = getGenomeWideAverage(Exons);\n",
      "                double genomeWideStdDev = getGenomeWideStdDev(Exons, genomeWideAve);\n",
      "                for (int i = 0; i < Exons.size(); i++) {\n",
      "                    Exon e = ExpectedValues.get(i);\n",
      "                    Exon s = StdDeviations.get(i);\n",
      "                    e.FPKM = genomeWideAve;\n",
      "                    s.FPKM = genomeWideStdDev;\n",
      "                }\n",
      "                System.out.println(\"Calculating States....\");\n",
      "                getStates(Exons, ExpectedValues, StdDeviations);\n",
      "                System.out.println(\"Finished calling\");\n",
      "            }\n",
      "        } catch (Exception e) {\n",
      "            System.err.println(\"Exception: \" + e.getMessage());\n",
      "            e.printStackTrace();\n",
      "        }\n",
      "    }\n",
      "\n",
      "    private static double getGenomeWideStdDev(ArrayList<Exon> exons, double avg) {\n",
      "        int total = 0;\n",
      "        for (Exon e : exons) {\n",
      "            total += (e.FPKM - avg) * (e.FPKM - avg);\n",
      "        }\n",
      "        return Math.sqrt(total / exons.size());\n",
      "    }\n",
      "\n",
      "    private static double getGenomeWideAverage(ArrayList<Exon> exons) {\n",
      "        int total = 0;\n",
      "        for (Exon e : exons) {\n",
      "            total += e.FPKM;\n",
      "        }\n",
      "        return total / exons.size();\n",
      "    }\n",
      "\n",
      "    private static void getStates(ArrayList<Exon> exons, ArrayList<Exon> expectedValues, ArrayList<Exon> stdDeviations) {\n",
      "        for (int i = 0; i < exons.size(); i++) {\n",
      "            Exon e = exons.get(i);\n",
      "            Exon mean = expectedValues.get(i);\n",
      "            Exon stddev = stdDeviations.get(i);\n",
      "            double normalPrior = 0.8;\n",
      "            double p[] = new double[5];\n",
      "            double p_normal = Normal(e.FPKM, mean.FPKM, stddev.FPKM);\n",
      "            p[0] = p_normal * normalPrior;\n",
      "            double p_homodel = Normal(e.FPKM, 0, stddev.FPKM);\n",
      "            p[1] = p_homodel * (1 - normalPrior) / 4;\n",
      "            double p_hetdel = Normal(e.FPKM, mean.FPKM * 0.5, stddev.FPKM);\n",
      "            p[2] = p_hetdel * (1 - normalPrior) / 4;\n",
      "            double p_amp1 = Normal(e.FPKM, mean.FPKM * 1.5, stddev.FPKM);\n",
      "            p[3] = p_amp1 * (1 - normalPrior) / 4;\n",
      "            double p_amp2 = Normal(e.FPKM, mean.FPKM * 2, stddev.FPKM);\n",
      "            p[4] = p_amp2 * (1 - normalPrior) / 4;\n",
      "            if (e.posLeft == 288236) System.out.println(\"f\");\n",
      "            double maxprob = Double.NEGATIVE_INFINITY;\n",
      "            int maxstate = -1;\n",
      "            for (int s = 0; s < 5; s++) {\n",
      "                if (p[s] > maxprob) {\n",
      "                    maxstate = s;\n",
      "                    maxprob = p[s];\n",
      "                }\n",
      "            }\n",
      "            if (p[0] >= maxprob) {\n",
      "                maxstate = 0;\n",
      "                maxprob = p[0];\n",
      "            }\n",
      "            e.state = HiddenMarkovModel.getStateFromIndex(maxstate);\n",
      "        }\n",
      "        if (outputFileName.equals(\"\")) {\n",
      "            for (int i = 0; i < Exons.size(); i++) {\n",
      "                Exon e = exons.get(i);\n",
      "                Exon mean = expectedValues.get(i);\n",
      "                Exon stddev = stdDeviations.get(i);\n",
      "                System.out.println(e + \"\\t\" + mean.FPKM + \"\\t\" + stddev.FPKM);\n",
      "            }\n",
      "        } else {\n",
      "            Writer output;\n",
      "            try {\n",
      "                output = new BufferedWriter(new FileWriter(outputFileName));\n",
      "                for (int i = 0; i < Exons.size(); i++) {\n",
      "                    Exon e = exons.get(i);\n",
      "                    Exon mean = expectedValues.get(i);\n",
      "                    Exon stddev = stdDeviations.get(i);\n",
      "                    output.write(e + \"\\t\" + mean.FPKM + \"\\t\" + stddev.FPKM + \"\\n\");\n",
      "                }\n",
      "                output.close();\n",
      "            } catch (IOException e1) {\n",
      "                e1.printStackTrace();\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "\n",
      "    private static double Normal(double x, double mean, double stddev) {\n",
      "        double var = Math.pow(stddev, 2);\n",
      "        double prob = Math.exp(-Math.pow(x - mean, 2) / (2 * var));\n",
      "        prob = prob / Math.sqrt(2 * Math.PI * var);\n",
      "        if (stddev == 0.0) if (x == mean) prob = 1; else prob = 0;\n",
      "        return prob;\n",
      "    }\n",
      "\n",
      "    private static void parseArguments(String arguments[]) {\n",
      "        try {\n",
      "            for (int index = 0; index < arguments.length; index++) {\n",
      "                String arg = arguments[index];\n",
      "                if (arg.equals(\"-e\")) exonFileName = arguments[index + 1]; else if (arg.equals(\"-t\")) printTimingMetrics = true; else if (arg.equals(\"-o\")) outputFileName = arguments[index + 1]; else if (arg.equals(\"-c\")) {\n",
      "                    String[] fields = arguments[index + 1].split(\"-\");\n",
      "                    for (int c = Integer.parseInt(fields[0]); c <= Integer.parseInt(fields[1]); c++) {\n",
      "                        chromosomes.add(c);\n",
      "                    }\n",
      "                } else if (arg.equals(\"-b\")) {\n",
      "                    String[] fields = arguments[index + 1].split(\",\");\n",
      "                    for (String f : fields) {\n",
      "                        baselineExonFileNames.add(f.trim());\n",
      "                    }\n",
      "                } else if (arg.equals(\"-p\")) {\n",
      "                    parameterFileName = arguments[index + 1].trim();\n",
      "                } else if (arg.equals(\"-threshold\")) {\n",
      "                    threshold = Double.parseDouble(arguments[index + 1].trim());\n",
      "                }\n",
      "            }\n",
      "            if (exonFileName.equals(\"\") || chromosomes.isEmpty() || parameterFileName.equals(\"\")) {\n",
      "                System.err.println(\"Improper Usage:\");\n",
      "                System.err.println(\"java CONDR -e <exonFileName> \" + \"-c <chromosomes (start-end)> \" + \"[-t] \" + \"[-o <output file name>]\");\n",
      "                System.exit(0);\n",
      "            }\n",
      "        } catch (Exception e) {\n",
      "            System.err.println(\"Exception: \" + e.getMessage());\n",
      "            e.printStackTrace();\n",
      "        }\n",
      "    }\n",
      "}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "c = get_dataset_generator(\n",
    "    dataset_name='bigclonebench',\n",
    "    mode='plagiarized',\n",
    "    **{}\n",
    ") \n",
    "\n",
    "nc = get_dataset_generator(\n",
    "    dataset_name='bigclonebench',\n",
    "    mode='non_plagiarized',\n",
    "    **{}\n",
    ")\n",
    "\n",
    "sg = selected_generator()\n",
    "\n",
    "\n",
    "cS = next(c)\n",
    "ncS = next(nc)\n",
    "\n",
    "sgS = next(sg)\n",
    "\n",
    "print(cS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "190998bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "import java.io.File;\n",
      "\n",
      "import javax.swing.JFileChooser;\n",
      "import javax.swing.JFrame;\n",
      "\n",
      "\n",
      "public class FileChooser {\n",
      "\t\n",
      "\tpublic static File chooseFileOpen(JFrame frame) {\n",
      "\t\tFile retval;\n",
      "\t\t\n",
      "\t\t//Create and configure file chooser\n",
      "\t\tJFileChooser fc = new JFileChooser();\n",
      "\t\tfc.setDialogTitle(\"Select input file.\");\n",
      "\t\tfc.setFileSelectionMode(JFileChooser.FILES_ONLY);\n",
      "\t\tfc.setMultiSelectionEnabled(false);\n",
      "\t\t\n",
      "\t\t//Show dialog and wait for user input\n",
      "\t\tint status = fc.showOpenDialog(frame);\n",
      "\t\t\n",
      "\t\t//React to input\n",
      "\t\tif(status == JFileChooser.APPROVE_OPTION) {\n",
      "\t\t\tretval = fc.getSelectedFile();\n",
      "\t\t} else if (status == JFileChooser.CANCEL_OPTION) {\n",
      "\t\t\tretval = null;\n",
      "\t\t} else {\n",
      "\t\t\tretval = null;\n",
      "\t\t}\n",
      "\t\t\n",
      "\t\t//Cleanup\n",
      "\t\tfc.setEnabled(false);\n",
      "\t\tfc.setVisible(false);\n",
      "\t\t\n",
      "\t\t//Return\n",
      "\t\treturn retval;\n",
      "\t}\n",
      "\t\n",
      "\tpublic static File chooseFileSave(JFrame frame) {\n",
      "\t\tFile retval;\n",
      "\t\t\n",
      "\t\t//Create and configure file chooser\n",
      "\t\tJFileChooser fc = new JFileChooser();\n",
      "\t\tfc.setDialogTitle(\"Select input file.\");\n",
      "\t\tfc.setFileSelectionMode(JFileChooser.FILES_ONLY);\n",
      "\t\tfc.setMultiSelectionEnabled(false);\n",
      "\t\t\n",
      "\t\t//Show dialog and wait for user input\n",
      "\t\tint status = fc.showSaveDialog(frame);\n",
      "\t\t\n",
      "\t\t//React to input\n",
      "\t\tif(status == JFileChooser.APPROVE_OPTION) {\n",
      "\t\t\tretval = fc.getSelectedFile();\n",
      "\t\t} else if (status == JFileChooser.CANCEL_OPTION) {\n",
      "\t\t\tretval = null;\n",
      "\t\t} else {\n",
      "\t\t\tretval = null;\n",
      "\t\t}\n",
      "\t\t\n",
      "\t\t//Cleanup\n",
      "\t\tfc.setEnabled(false);\n",
      "\t\tfc.setVisible(false);\n",
      "\t\t\n",
      "\t\t//Return\n",
      "\t\treturn retval;\n",
      "\t}\n",
      "\t\n",
      "\tpublic static File[] chooseFileOpenMultiple(JFrame frame) {\n",
      "\t\tFile retval[];\n",
      "\t\t\n",
      "\t\t//Create and configure file chooser\n",
      "\t\tJFileChooser fc = new JFileChooser();\n",
      "\t\tfc.setDialogTitle(\"Select input file.\");\n",
      "\t\tfc.setFileSelectionMode(JFileChooser.FILES_ONLY);\n",
      "\t\tfc.setMultiSelectionEnabled(true);\n",
      "\t\t\n",
      "\t\t//Show dialog and wait for user input\n",
      "\t\tint status = fc.showSaveDialog(frame);\n",
      "\t\t\n",
      "\t\t//React to input\n",
      "\t\tif(status == JFileChooser.APPROVE_OPTION) {\n",
      "\t\t\tretval = fc.getSelectedFiles();\n",
      "\t\t} else if (status == JFileChooser.CANCEL_OPTION) {\n",
      "\t\t\tretval = null;\n",
      "\t\t} else {\n",
      "\t\t\tretval = null;\n",
      "\t\t}\n",
      "\t\t\n",
      "\t\t//Cleanup\n",
      "\t\tfc.setEnabled(false);\n",
      "\t\tfc.setVisible(false);\n",
      "\t\t\n",
      "\t\t//Return\n",
      "\t\treturn retval;\n",
      "\t}\n",
      "\t\n",
      "\tpublic static File[] chooseFileDirectory(JFrame frame) {\n",
      "\t\tFile retval[];\n",
      "\t\t\n",
      "\t\t//Create and configure file chooser\n",
      "\t\tJFileChooser fc = new JFileChooser();\n",
      "\t\tfc.setDialogTitle(\"Select input file.\");\n",
      "\t\tfc.setFileSelectionMode(JFileChooser.DIRECTORIES_ONLY);\n",
      "\t\tfc.setMultiSelectionEnabled(false);\n",
      "\t\t\n",
      "\t\t//Show dialog and wait for user input\n",
      "\t\tint status = fc.showSaveDialog(frame);\n",
      "\t\t\n",
      "\t\t//React to input\n",
      "\t\tif(status == JFileChooser.APPROVE_OPTION) {\n",
      "\t\t\tretval = fc.getSelectedFiles();\n",
      "\t\t} else if (status == JFileChooser.CANCEL_OPTION) {\n",
      "\t\t\tretval = null;\n",
      "\t\t} else {\n",
      "\t\t\tretval = null;\n",
      "\t\t}\n",
      "\t\t\n",
      "\t\t//Cleanup\n",
      "\t\tfc.setEnabled(false);\n",
      "\t\tfc.setVisible(false);\n",
      "\t\t\n",
      "\t\t//Return\n",
      "\t\treturn retval;\n",
      "\t}\n",
      "\t\n",
      "\t//public static void main(String args[]) {\n",
      "\t//\tJFrame frame = new JFrame();\n",
      "\t//\tframe.setVisible(true);\n",
      "\t//\tSystem.out.println(FileChooser.chooseFileOpen(frame));\n",
      "\t//\tSystem.out.println(FileChooser.chooseFileSave(frame));\n",
      "\t//\tSystem.out.println(FileChooser.chooseFileOpenMultiple(frame));\n",
      "\t//\tframe.setVisible(false);\n",
      "\t//\tframe.disable();\n",
      "\t//}\n",
      "}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(ncS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "594211c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "package org.communityca.dao.spring;\n",
      "\n",
      "import org.springframework.beans.MutablePropertyValues;\n",
      "import org.springframework.beans.PropertyValue;\n",
      "import org.springframework.beans.factory.NoSuchBeanDefinitionException;\n",
      "import org.springframework.beans.factory.config.BeanDefinition;\n",
      "import org.springframework.beans.factory.config.BeanFactoryPostProcessor;\n",
      "import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;\n",
      "import java.util.ArrayList;\n",
      "import java.util.List;\n",
      "import java.util.Properties;\n",
      "\n",
      "/**\n",
      " * <p>Adds Hibernate persistent class definitions to an existing Spring Session Factory bean, possibly defined\n",
      " * within a seperate Spring configuration file in a seperate jar file. By using this extension factory developers can\n",
      " * add persistent classes to an AppFuse application without modifying any of the existing AppFuse Spring configuration\n",
      " * or jar distribution files.\n",
      " * \n",
      " * <p>As an example consider the following Spring bean configuration:\n",
      " * \n",
      " * <pre>\n",
      " * &lt;bean class=&quot;org.communityca.dao.spring.HibernateExtensionPostProcessor&quot;&gt;\n",
      " *     &lt;property name=&quot;mappingResources&quot;&gt;\n",
      " *         &lt;list&gt;\n",
      " *             &lt;value&gt;org/communityca/model/Foo.hbm.xml&lt;/value&gt;\n",
      " *         &lt;/list&gt;\n",
      " *     &lt;/property&gt;\n",
      " *     &lt;property name=&quot;annotatedClasses&quot;&gt;\n",
      " *         &lt;list&gt;\n",
      " *             &lt;value&gt;org.communityca.model.Bar&lt;/value&gt;\n",
      " *         &lt;/list&gt;\n",
      " *     &lt;/property&gt;\n",
      " * &lt;/bean&gt;\n",
      " * </pre>\n",
      " * \n",
      " * <p>The snippet will add two persistent classes to an existing Session Factory bean called &quot;sessionFactory&quot;.\n",
      " * Note that the extension can handle both annotated classes and the more traditional .hbm.xml files. Assuming that\n",
      " * these persistent classes are packaged in a jar called extension.jar which contains the Spring configuration file\n",
      " * applicationContext-dao.xml at the root level, then the standard AppFuse configuration will automatically pick up the\n",
      " * new Spring configuration file and the new persistent classes will be added to the list already defined for the\n",
      " * session factory bean configured within the standard appfuse-hibernate.jar file. And all this without needing to\n",
      " * touch the original AppFuse configuration files!\n",
      " *\n",
      " * @author Michael Horwitz\n",
      " */\n",
      "public class HibernateExtensionPostProcessor implements BeanFactoryPostProcessor {\n",
      "\n",
      "    private String sessionFactoryBeanName = \"sessionFactory\";\n",
      "\n",
      "    private List mappingResources;\n",
      "\n",
      "    private List annotatedClasses;\n",
      "\n",
      "    private List configLocations;\n",
      "\n",
      "    private Properties hibernateProperties;\n",
      "\n",
      "    /**\n",
      "     * Adds the annotated classes and the mapping resources to the existing Session Factory configuration.\n",
      "     * @param configurableListableBeanFactory the good ol' bean factory\n",
      "     */\n",
      "    @SuppressWarnings(\"unchecked\")\n",
      "    public void postProcessBeanFactory(ConfigurableListableBeanFactory configurableListableBeanFactory) {\n",
      "        if (configurableListableBeanFactory.containsBean(sessionFactoryBeanName)) {\n",
      "            BeanDefinition sessionFactoryBeanDefinition = configurableListableBeanFactory.getBeanDefinition(sessionFactoryBeanName);\n",
      "            MutablePropertyValues propertyValues = sessionFactoryBeanDefinition.getPropertyValues();\n",
      "            if (mappingResources != null) {\n",
      "                PropertyValue propertyValue = propertyValues.getPropertyValue(\"mappingResources\");\n",
      "                if (propertyValue == null) {\n",
      "                    propertyValue = new PropertyValue(\"mappingResources\", new ArrayList());\n",
      "                    propertyValues.addPropertyValue(propertyValue);\n",
      "                }\n",
      "                List existingMappingResources = (List) propertyValue.getValue();\n",
      "                existingMappingResources.addAll(mappingResources);\n",
      "            }\n",
      "            if (annotatedClasses != null) {\n",
      "                PropertyValue propertyValue = propertyValues.getPropertyValue(\"annotatedClasses\");\n",
      "                if (propertyValue == null) {\n",
      "                    propertyValue = new PropertyValue(\"annotatedClasses\", new ArrayList());\n",
      "                    propertyValues.addPropertyValue(propertyValue);\n",
      "                }\n",
      "                List existingMappingResources = (List) propertyValue.getValue();\n",
      "                existingMappingResources.addAll(annotatedClasses);\n",
      "            }\n",
      "            if (configLocations != null) {\n",
      "                PropertyValue propertyValue = propertyValues.getPropertyValue(\"configLocations\");\n",
      "                if (propertyValue == null) {\n",
      "                    propertyValue = new PropertyValue(\"configLocations\", new ArrayList());\n",
      "                    propertyValues.addPropertyValue(propertyValue);\n",
      "                }\n",
      "                List existingConfigLocations = (List) propertyValue.getValue();\n",
      "                existingConfigLocations.addAll(configLocations);\n",
      "            }\n",
      "            if (hibernateProperties != null) {\n",
      "                PropertyValue propertyValue = propertyValues.getPropertyValue(\"hibernateProperties\");\n",
      "                if (propertyValue == null) {\n",
      "                    propertyValue = new PropertyValue(\"hibernateProperties\", new Properties());\n",
      "                    propertyValues.addPropertyValue(propertyValue);\n",
      "                }\n",
      "                Properties existingHibernateProperties = (Properties) propertyValue.getValue();\n",
      "                existingHibernateProperties.putAll(hibernateProperties);\n",
      "            }\n",
      "        } else {\n",
      "            throw new NoSuchBeanDefinitionException(\"No bean named [\" + sessionFactoryBeanName + \"] exists within the bean factory. \" + \"Cannot post process session factory to add Hibernate resource definitions.\");\n",
      "        }\n",
      "    }\n",
      "\n",
      "    /**\n",
      "     * Set the name of the SessionFactory bean. By default this post processor looks for a\n",
      "     * bean of name &quot;sessionFactory&quot;\n",
      "     *\n",
      "     * @param sessionFactoryBeanName The name of the session factory bean.\n",
      "     */\n",
      "    public void setSessionFactoryBeanName(String sessionFactoryBeanName) {\n",
      "        this.sessionFactoryBeanName = sessionFactoryBeanName;\n",
      "    }\n",
      "\n",
      "    /**\n",
      "     * Set the list of mapping resources (.hbm.xml files) to be added to the session factory.\n",
      "     *\n",
      "     * @param mappingResources The list of mapping resources.\n",
      "     */\n",
      "    public void setMappingResources(List mappingResources) {\n",
      "        this.mappingResources = mappingResources;\n",
      "    }\n",
      "\n",
      "    /**\n",
      "     * The list of annotated classes to add to the session factory.\n",
      "     *\n",
      "     * @param annotatedClasses The list of annotated classes that need to be added.\n",
      "     */\n",
      "    public void setAnnotatedClasses(List annotatedClasses) {\n",
      "        this.annotatedClasses = annotatedClasses;\n",
      "    }\n",
      "\n",
      "    /**\n",
      "     * The list of configuration locations (i.e. classpath:hibernate.cfg.xml) to add to the session factory\n",
      "     *\n",
      "     * @param configLocations The list of configuration locations that need to be added.\n",
      "     */\n",
      "    public void setConfigLocations(List configLocations) {\n",
      "        this.configLocations = configLocations;\n",
      "    }\n",
      "\n",
      "    /**\n",
      "     * Hibernate properties to add to the session factory.\n",
      "     *\n",
      "     * @param hibernateProperties The list of additional properties.\n",
      "     */\n",
      "    public void setHibernateProperties(Properties hibernateProperties) {\n",
      "        this.hibernateProperties = hibernateProperties;\n",
      "    }\n",
      "}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(sgS)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
